core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterManager.java

```diff
@@ -0,0 +1,102 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class ScaleWriterManager
+{
+    private static final Logger log = Logger.get(ScaleWriterExchanger.class);
+
+    private final LocalExchangeMemoryManager memoryManager;
+    private final int totalWriters;
+    private final long bufferSizeMinScalingThreshold;
+    private final Supplier<Long> physicalWrittenBytesSupplier;
+    private final long writerMinSize;
+    private final Runnable setWritersFullyScaled;
+    private final AtomicBoolean isSetWritersFullyScaledExecuted = new AtomicBoolean(false);
+
+    // start with single writer and increase the writer count based on
+    // physical written bytes and buffer utilization.
+    @GuardedBy("this")
+    private int writerCount = 1;
+
+    @GuardedBy("this")
+    private long lastScaleUpPhysicalWrittenBytes;
+
+    public ScaleWriterManager(
+            LocalExchangeMemoryManager memoryManager,
+            int totalWriters,
+            long maxBufferedBytes,
+            Supplier<Long> physicalWrittenBytesSupplier,
+            DataSize writerMinSize,
+            Runnable setWritersFullyScaled)
+    {
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.totalWriters = totalWriters;
+        // We are aiming to keep the buffer size below 50% of its limit
+        this.bufferSizeMinScalingThreshold = maxBufferedBytes / 2;
+        this.physicalWrittenBytesSupplier = requireNonNull(physicalWrittenBytesSupplier, "physicalWrittenBytesSupplier is null");
+        this.writerMinSize = requireNonNull(writerMinSize, "writerMinSize is null").toBytes();
+        this.setWritersFullyScaled = requireNonNull(setWritersFullyScaled, "setWritersFullyScaled is null");
+    }
+
+    public int getWriterCount()
+    {
+        updateWriterCount(memoryManager.getBufferedBytes());
+        return writerCount;
+    }
+
+    private void updateWriterCount(long bufferedBytes)
+    {
+        // Scale up writers when current buffer memory utilization is more than 50% of the
+        // maximum and physical written bytes by the last scaled up writer is greater than
+        // writerMinSize.
+        if (writerCount < totalWriters && bufferedBytes >= bufferSizeMinScalingThreshold) {
+            synchronized (this) {
```

Comment:
Could you explain the need for locking here ?
I think we get here from LocalExchangeSinkOperator#addInput and that's single threaded AFAIK

Reply:
https://github.com/trinodb/trino/pull/13111#discussion_r932436698

Because of the above issue, we have to use a single instance of ScaleWriterManager across all exchangers.